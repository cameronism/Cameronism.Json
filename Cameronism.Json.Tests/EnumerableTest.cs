using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Cameronism.Json;
using Xunit;

namespace Cameronism.Json.Tests
{
	public class EnumerableTest
	{
		class ExplicitEnumerableBadCount : IEnumerable<bool>
		{
			IEnumerator<bool> IEnumerable<bool>.GetEnumerator() { throw new NotImplementedException(); } 
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); } 
			/// <summary>
			/// This should not be picked up since neither ICollection<> or IReadOnlyCollection are implemented
			/// </summary>
			public int Count
			{
				get
				{
					return 0;
				}
			}
		}

		class ExplicitReadOnlyCollection : IReadOnlyCollection<Guid>
		{
			int IReadOnlyCollection<Guid>.Count { get { throw new NotImplementedException(); } } 
			IEnumerator<Guid> IEnumerable<Guid>.GetEnumerator() { throw new NotImplementedException(); } 
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
		}
		class ExplicitReadOnlyCollectionImplicitCount : IReadOnlyCollection<byte>
		{
			public int Count { get { throw new NotImplementedException(); } }
			IEnumerator<byte> IEnumerable<byte>.GetEnumerator() { throw new NotImplementedException(); }
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
		}
		class ExplicitReadOnlyCollectionPublicEnumerator : IReadOnlyCollection<sbyte>
		{
			public struct Enumerator
			{
				public bool MoveNext() { return false; }
				public sbyte Current { get { return 0; } }
			}

			public Enumerator GetEnumerator() { return new Enumerator(); }
			int IReadOnlyCollection<sbyte>.Count { get { throw new NotImplementedException(); } } 
			IEnumerator<sbyte> IEnumerable<sbyte>.GetEnumerator() { throw new NotImplementedException(); } 
			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
		}

		class MyReadOnlyDictionary<TKey, TValue> : IReadOnlyDictionary<TKey, TValue>
		{
			KeyValuePair<TKey,TValue>[] _Pairs;

			private MyReadOnlyDictionary(IEnumerable<KeyValuePair<TKey, TValue>> pairs)
			{
				_Pairs = pairs.ToArray();
			}

			public bool ContainsKey(TKey key)
			{
				throw new NotImplementedException();
			}

			public IEnumerable<TKey> Keys
			{
				get { throw new NotImplementedException(); }
			}

			public bool TryGetValue(TKey key, out TValue value)
			{
				throw new NotImplementedException();
			}

			public IEnumerable<TValue> Values
			{
				get { throw new NotImplementedException(); }
			}

			public TValue this[TKey key]
			{
				get { throw new NotImplementedException(); }
			}

			public int Count
			{
				get { throw new NotImplementedException(); }
			}

			public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
			{
				return ((IEnumerable<KeyValuePair<TKey, TValue>>)_Pairs).GetEnumerator();
			}

			System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
			{
				return _Pairs.GetEnumerator();
			}

			public static MyReadOnlyDictionary<TKey, TValue> Create(IEnumerable<KeyValuePair<TKey, TValue>> pairs)
			{
				return new MyReadOnlyDictionary<TKey, TValue>(pairs);
			}
		}



		[Fact]
		public void FindMethods()
		{
			var sb = new StringBuilder();

			foreach (var type in new[]
			{
				typeof(List<int>),
				typeof(IEnumerable<int>),
				typeof(ExplicitEnumerableBadCount),
				typeof(ExplicitReadOnlyCollection),
				typeof(ExplicitReadOnlyCollectionImplicitCount),
				typeof(ExplicitReadOnlyCollectionPublicEnumerator),
				typeof(ICollection<short>),
				typeof(IReadOnlyCollection<ushort>),
				typeof(IReadOnlyList<long>),
				typeof(IList<ulong>),
				this.GetType(), // throw in something non enumerable
			})
			{
				var methods = EnumerableInfo.FindMethods(type);
				sb.AppendLine("# " + SchemaTest.HumanName(type));
				DescribeMethod(sb, () => methods.GetEnumerator);
				DescribeMethod(sb, () => methods.MoveNext);
				DescribeMethod(sb, () => methods.get_Current);
				DescribeMethod(sb, () => methods.get_Count);
				DescribeMethod(sb, () => methods.Dispose);

				sb.AppendLine();
			}

			ApprovalTests.Approvals.Verify(sb.ToString());
		}

		static void DescribeMethod(StringBuilder sb, Expression<Func<MethodInfo>> fmi)
		{
			var prop = fmi.Body as MemberExpression;
			sb.AppendLine("## " + prop.Member.Name);

			var mi = fmi.Compile().Invoke();
			if (mi == null)
			{
				sb.AppendLine("`null`");
			}
			else
			{
				sb.AppendLine("Return Type: " + SchemaTest.HumanName(mi.ReturnType));
				sb.AppendLine("IsVirtual: " + mi.IsVirtual);
				sb.AppendLine("Declaring Type: " + SchemaTest.HumanName(mi.DeclaringType));
				foreach (var arg in mi.GetParameters())
				{
					sb.AppendLine(String.Format("{0}: {1}", arg.Name, SchemaTest.HumanName(arg.ParameterType)));
				}
			}

			sb.AppendLine();
		}

		[Fact]
		public void SerializeEnumerable()
		{
			var buffer = new byte[4096];
			var sb = new StringBuilder();
			int failed = 0;

			failed += SerializeValues<IEnumerable<int>>(sb, buffer, true,
				null, 
				new List<int> { },
				new List<int> { 1 },
				new List<int> { 1, 2 },
				new int[] { },
				new int[] { 2 },
				new int[] { 2, 4 },
				Enumerable.Empty<int>(),
				Enumerable.Repeat(3, 0),
				Enumerable.Repeat(3, 1),
				Enumerable.Repeat(3, 2),
				Enumerable.Repeat(3, 3),
				null);

			failed += SerializeValues(sb, buffer, true,
				new List<int> { },
				new List<int> { 1 },
				new List<int> { 1, 2 },
				new List<int> { 1, 2, 3 });

			failed += SerializeValues<IList<int?>>(sb, buffer, false,
				new int?[] { },
				new List<int?> { null },
				new int?[] { 1 },
				new List<int?> { 1, null },
				new int?[] { 1, 2 },
				new List<int?> { 1, 2, null },
				new int?[] { 1, 2, 3 },
				new List<int?> { 1, 2, 3, null });

			failed += SerializeValues<IReadOnlyCollection<int?>>(sb, buffer, false,
				new int?[] { },
				new List<int?> { null }.AsReadOnly(),
				new int?[] { 1 },
				new List<int?> { 1, null },
				new int?[] { 1, 2 },
				new List<int?> { 1, 2, null },
				new int?[] { 1, 2, 3 },
				new List<int?> { 1, 2, 3, null });


			failed += SerializeValues(sb, buffer, false,
				new List<int?> { null }.AsReadOnly());

			// Newtonsoft thinks this should be an object not an array :(
			//failed += SerializeValues<IEnumerable<KeyValuePair<string, int>>>(sb, buffer, true,
			//	new Dictionary<string, int> { },
			//	new Dictionary<string, int> { { "0", 0 } },
			//	new Dictionary<string, int> { { "0", 0 }, { "1", 1 } },
			//	null);

			failed += SerializeValues(sb, buffer, true,
				new Dictionary<string, int> { },
				new Dictionary<string, int> { { "0", 0 } },
				new Dictionary<string, int> { { "0", 0 }, { "1", 1 } },
				null);

			failed += SerializeValues(sb, buffer, false,
				MyReadOnlyDictionary<string, int>.Create(new Dictionary<string, int> { }),
				MyReadOnlyDictionary<string, int>.Create(new Dictionary<string, int> { { "0", 0 } }),
				MyReadOnlyDictionary<string, int>.Create(new Dictionary<string, int> { { "0", 0 }, { "1", 1 } }),
				null);

			failed += SerializeValues(sb, buffer, false,
				new Dictionary<char, int> { },
				new Dictionary<char, int> { { '0', 0 } },
				new Dictionary<char, int> { { '0', 0 }, { '1', 1 } },
				null);

			failed += SerializeValues(sb, buffer, false,
				new Dictionary<Guid, int[]> { },
				new Dictionary<Guid, int[]> { { Guid.Empty, null } },
				new Dictionary<Guid, int[]> { { Guid.Empty, new[] { 0 } }, { Guid.Parse("838f5c03-ed56-41e1-9264-686bb18d0a77"), new[] { 1, 2 } } },
				null);

			ApprovalTests.Approvals.Verify(sb.ToString());
			Assert.True(failed == 0, "Look at the approval " + failed + " tests failed");
		}

		static int SerializeValues<T>(StringBuilder sb, byte[] buffer, bool approveIL, params T[] values)
		{
			return DelegateBuilderTest.SerializeValues(sb, buffer, values, approveIL);
		}

		class MethodBucket
		{
			public int _1 () { return 0; }
			public string _2 () { return null; }
			public IDisposable _3() { return null; }
			private int _4() { return 0; }
			private IDisposable _5() { return null; }
		}

		[Fact]
		public void VerifyGetEnumeratorOrder()
		{
			// repeat it several times so that we try both directions of all comparisons
			var methods = Enumerable.Repeat(typeof(MethodBucket).GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly), 16)
				.SelectMany(xs => xs)
				.ToArray();

			var expectedOrder = methods.OrderBy(mi => int.Parse(mi.Name.Substring(1))).ToList();
			Array.Sort(methods, EnumerableInfo.SortEnumeratorCandidates);

			Assert.Equal(expectedOrder.Select(mi => mi.Name), methods.Select(mi => mi.Name), StringComparer.Ordinal);
		}
	}
}
