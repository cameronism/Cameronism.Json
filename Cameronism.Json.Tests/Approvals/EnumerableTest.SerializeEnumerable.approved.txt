

# System.Collections.Generic.IEnumerable<System.Int32>

## il
ldarg.1
stloc.0
ldarg.2
stloc.1
ldloc.1
ldarg.0
ldind.ref
dup
brtrue.s ifNotNull_0
pop
pop
ldloc.1
ldc.i4.4
bge.s enoughAvailable_2
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_2:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 110
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 117
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 108
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 108
stind.i1
ldloc.1
ldc.i4.4
sub
stloc.1
ldc.i4.4
br ifNull_1

ifNotNull_0:
ldloc.1
ldc.i4.1
bge.s enoughAvailable_3
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_3:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 91
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
callvirt System.Collections.Generic.IEnumerator`1[System.Int32] GetEnumerator()
dup
callvirt Boolean MoveNext()
brfalse.s closeArray_6
dup
callvirt Int32 get_Current()
ldloc.0
ldloc.1
call Int32 WriteInt32(Int32, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_7
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_7:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1
br.s loopBottom_4

loopTop_5:
ldloc.1
ldc.i4.1
bge.s enoughAvailable_8
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_8:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 44
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
dup
callvirt Int32 get_Current()
ldloc.0
ldloc.1
call Int32 WriteInt32(Int32, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_9
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_9:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1

loopBottom_4:
dup
callvirt Boolean MoveNext()
brtrue.s loopTop_5

closeArray_6:
pop
ldloc.1
ldc.i4.1
bge.s enoughAvailable_10
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_10:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 93
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
ldloc.1
sub

ifNull_1:
ret




## Newtonsoft 0


0000   6e 75 6c 6c                                         null

## Cameronism.Json 0


0000   6e 75 6c 6c                                         null

### Equal: True

## Newtonsoft 1


0000   5b 5d                                               []

## Cameronism.Json 1


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 2


0000   5b 31 5d                                            [1]

## Cameronism.Json 2


0000   5b 31 5d                                            [1]

### Equal: True

## Newtonsoft 3


0000   5b 31 2c 32 5d                                      [1,2]

## Cameronism.Json 3


0000   5b 31 2c 32 5d                                      [1,2]

### Equal: True

## Newtonsoft 4


0000   5b 5d                                               []

## Cameronism.Json 4


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 5


0000   5b 32 5d                                            [2]

## Cameronism.Json 5


0000   5b 32 5d                                            [2]

### Equal: True

## Newtonsoft 6


0000   5b 32 2c 34 5d                                      [2,4]

## Cameronism.Json 6


0000   5b 32 2c 34 5d                                      [2,4]

### Equal: True

## Newtonsoft 7


0000   5b 5d                                               []

## Cameronism.Json 7


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 8


0000   5b 5d                                               []

## Cameronism.Json 8


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 9


0000   5b 33 5d                                            [3]

## Cameronism.Json 9


0000   5b 33 5d                                            [3]

### Equal: True

## Newtonsoft 10


0000   5b 33 2c 33 5d                                      [3,3]

## Cameronism.Json 10


0000   5b 33 2c 33 5d                                      [3,3]

### Equal: True

## Newtonsoft 11


0000   5b 33 2c 33 2c 33 5d                                [3,3,3]

## Cameronism.Json 11


0000   5b 33 2c 33 2c 33 5d                                [3,3,3]

### Equal: True

## Newtonsoft 12


0000   6e 75 6c 6c                                         null

## Cameronism.Json 12


0000   6e 75 6c 6c                                         null

### Equal: True

# System.Collections.Generic.List<System.Int32>

## il
ldarg.1
stloc.0
ldarg.2
stloc.1
ldloc.1
ldarg.0
ldind.ref
dup
brtrue.s ifNotNull_0
pop
pop
ldloc.1
ldc.i4.4
bge.s enoughAvailable_2
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_2:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 110
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 117
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 108
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 108
stind.i1
ldloc.1
ldc.i4.4
sub
stloc.1
ldc.i4.4
br ifNull_1

ifNotNull_0:
ldloc.1
ldc.i4.1
bge.s enoughAvailable_3
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_3:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 91
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
call Enumerator GetEnumerator()
stloc.2
ldloca.s 2
dup
constrained. System.Collections.Generic.List`1+Enumerator[System.Int32]
callvirt Boolean MoveNext()
brfalse closeArray_6
dup
constrained. System.Collections.Generic.List`1+Enumerator[System.Int32]
callvirt Int32 get_Current()
ldloc.0
ldloc.1
call Int32 WriteInt32(Int32, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_8
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_8:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1
br.s loopBottom_4

loopTop_5:
ldloc.1
ldc.i4.1
bge.s enoughAvailable_9
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_9:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 44
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
dup
constrained. System.Collections.Generic.List`1+Enumerator[System.Int32]
callvirt Int32 get_Current()
ldloc.0
ldloc.1
call Int32 WriteInt32(Int32, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_10
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_10:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1

loopBottom_4:
dup
constrained. System.Collections.Generic.List`1+Enumerator[System.Int32]
callvirt Boolean MoveNext()
brtrue.s loopTop_5

closeArray_6:
pop
ldloc.1
ldc.i4.1
bge.s enoughAvailable_11
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_11:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 93
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
ldloc.1
sub

ifNull_1:
ret




## Newtonsoft 0


0000   5b 5d                                               []

## Cameronism.Json 0


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 1


0000   5b 31 5d                                            [1]

## Cameronism.Json 1


0000   5b 31 5d                                            [1]

### Equal: True

## Newtonsoft 2


0000   5b 31 2c 32 5d                                      [1,2]

## Cameronism.Json 2


0000   5b 31 2c 32 5d                                      [1,2]

### Equal: True

## Newtonsoft 3


0000   5b 31 2c 32 2c 33 5d                                [1,2,3]

## Cameronism.Json 3


0000   5b 31 2c 32 2c 33 5d                                [1,2,3]

### Equal: True

# System.Collections.Generic.IList<System.Nullable<System.Int32>>


## Newtonsoft 0


0000   5b 5d                                               []

## Cameronism.Json 0


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 1


0000   5b 6e 75 6c 6c 5d                                   [null]

## Cameronism.Json 1


0000   5b 6e 75 6c 6c 5d                                   [null]

### Equal: True

## Newtonsoft 2


0000   5b 31 5d                                            [1]

## Cameronism.Json 2


0000   5b 31 5d                                            [1]

### Equal: True

## Newtonsoft 3


0000   5b 31 2c 6e 75 6c 6c 5d                             [1,null]

## Cameronism.Json 3


0000   5b 31 2c 6e 75 6c 6c 5d                             [1,null]

### Equal: True

## Newtonsoft 4


0000   5b 31 2c 32 5d                                      [1,2]

## Cameronism.Json 4


0000   5b 31 2c 32 5d                                      [1,2]

### Equal: True

## Newtonsoft 5


0000   5b 31 2c 32 2c 6e 75 6c   6c 5d                     [1,2,nul l]

## Cameronism.Json 5


0000   5b 31 2c 32 2c 6e 75 6c   6c 5d                     [1,2,nul l]

### Equal: True

## Newtonsoft 6


0000   5b 31 2c 32 2c 33 5d                                [1,2,3]

## Cameronism.Json 6


0000   5b 31 2c 32 2c 33 5d                                [1,2,3]

### Equal: True

## Newtonsoft 7


0000   5b 31 2c 32 2c 33 2c 6e   75 6c 6c 5d               [1,2,3,n ull]

## Cameronism.Json 7


0000   5b 31 2c 32 2c 33 2c 6e   75 6c 6c 5d               [1,2,3,n ull]

### Equal: True

# System.Collections.Generic.IReadOnlyCollection<System.Nullable<System.Int32>>


## Newtonsoft 0


0000   5b 5d                                               []

## Cameronism.Json 0


0000   5b 5d                                               []

### Equal: True

## Newtonsoft 1


0000   5b 6e 75 6c 6c 5d                                   [null]

## Cameronism.Json 1


0000   5b 6e 75 6c 6c 5d                                   [null]

### Equal: True

## Newtonsoft 2


0000   5b 31 5d                                            [1]

## Cameronism.Json 2


0000   5b 31 5d                                            [1]

### Equal: True

## Newtonsoft 3


0000   5b 31 2c 6e 75 6c 6c 5d                             [1,null]

## Cameronism.Json 3


0000   5b 31 2c 6e 75 6c 6c 5d                             [1,null]

### Equal: True

## Newtonsoft 4


0000   5b 31 2c 32 5d                                      [1,2]

## Cameronism.Json 4


0000   5b 31 2c 32 5d                                      [1,2]

### Equal: True

## Newtonsoft 5


0000   5b 31 2c 32 2c 6e 75 6c   6c 5d                     [1,2,nul l]

## Cameronism.Json 5


0000   5b 31 2c 32 2c 6e 75 6c   6c 5d                     [1,2,nul l]

### Equal: True

## Newtonsoft 6


0000   5b 31 2c 32 2c 33 5d                                [1,2,3]

## Cameronism.Json 6


0000   5b 31 2c 32 2c 33 5d                                [1,2,3]

### Equal: True

## Newtonsoft 7


0000   5b 31 2c 32 2c 33 2c 6e   75 6c 6c 5d               [1,2,3,n ull]

## Cameronism.Json 7


0000   5b 31 2c 32 2c 33 2c 6e   75 6c 6c 5d               [1,2,3,n ull]

### Equal: True

# System.Collections.ObjectModel.ReadOnlyCollection<System.Nullable<System.Int32>>


## Newtonsoft 0


0000   5b 6e 75 6c 6c 5d                                   [null]

## Cameronism.Json 0


0000   5b 6e 75 6c 6c 5d                                   [null]

### Equal: True

# System.Collections.Generic.Dictionary<System.String, System.Int32>

## il
ldarg.1
stloc.0
ldarg.2
stloc.1
ldloc.1
ldarg.0
ldind.ref
dup
brtrue.s ifNotNull_0
pop
pop
ldloc.1
ldc.i4.4
bge.s enoughAvailable_2
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_2:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 110
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 117
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 108
stind.i1
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 108
stind.i1
ldloc.1
ldc.i4.4
sub
stloc.1
ldc.i4.4
br ifNull_1

ifNotNull_0:
ldloc.1
ldc.i4.1
bge.s enoughAvailable_3
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_3:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 123
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
call Enumerator GetEnumerator()
stloc.2
ldloca.s 2
dup
constrained. System.Collections.Generic.Dictionary`2+Enumerator[System.String,System.Int32]
callvirt Boolean MoveNext()
brfalse closeArray_6
dup
constrained. System.Collections.Generic.Dictionary`2+Enumerator[System.String,System.Int32]
callvirt System.Collections.Generic.KeyValuePair`2[System.String,System.Int32] get_Current()
stloc.3
ldloca.s 3
dup
call System.String get_Key()
ldloc.0
ldloc.1
call Int32 WriteStringUtf8(System.String, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_9
pop
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_9:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1
ldloc.1
ldc.i4.1
bge.s enoughAvailable_10
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_10:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 58
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
call Int32 get_Value()
ldloc.0
ldloc.1
call Int32 WriteInt32(Int32, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_11
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_11:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1
br loopBottom_4

loopTop_5:
ldloc.1
ldc.i4.1
bge.s enoughAvailable_12
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_12:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 44
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
dup
constrained. System.Collections.Generic.Dictionary`2+Enumerator[System.String,System.Int32]
callvirt System.Collections.Generic.KeyValuePair`2[System.String,System.Int32] get_Current()
stloc.3
ldloca.s 3
dup
call System.String get_Key()
ldloc.0
ldloc.1
call Int32 WriteStringUtf8(System.String, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_14
pop
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_14:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1
ldloc.1
ldc.i4.1
bge.s enoughAvailable_15
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_15:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 58
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
call Int32 get_Value()
ldloc.0
ldloc.1
call Int32 WriteInt32(Int32, Byte*, Int32)
dup
ldc.i4.0
bgt.s success_16
pop
pop
pop
ldarg.2
ldc.i4.s 254
mul
ret

success_16:
dup
conv.i
ldloc.0
add
stloc.0
ldc.i4.m1
mul
ldloc.1
add
stloc.1

loopBottom_4:
dup
constrained. System.Collections.Generic.Dictionary`2+Enumerator[System.String,System.Int32]
callvirt Boolean MoveNext()
brtrue loopTop_5

closeArray_6:
pop
ldloc.1
ldc.i4.1
bge.s enoughAvailable_17
pop
ldarg.2
ldc.i4.s 254
mul
ret

enoughAvailable_17:
ldloc.0
dup
ldc.i4.1
conv.i
add
stloc.0
ldc.i4.s 125
stind.i1
ldloc.1
ldc.i4.1
sub
stloc.1
ldloc.1
sub

ifNull_1:
ret




## Newtonsoft 0


0000   7b 7d                                               {}

## Cameronism.Json 0


0000   7b 7d                                               {}

### Equal: True

## Newtonsoft 1


0000   7b 22 30 22 3a 30 7d                                {"0":0}

## Cameronism.Json 1


0000   7b 22 30 22 3a 30 7d                                {"0":0}

### Equal: True

## Newtonsoft 2


0000   7b 22 30 22 3a 30 2c 22   31 22 3a 31 7d            {"0":0," 1":1}

## Cameronism.Json 2


0000   7b 22 30 22 3a 30 2c 22   31 22 3a 31 7d            {"0":0," 1":1}

### Equal: True

## Newtonsoft 3


0000   6e 75 6c 6c                                         null

## Cameronism.Json 3


0000   6e 75 6c 6c                                         null

### Equal: True

# Cameronism.Json.Tests.EnumerableTest+MyReadOnlyDictionary<System.String, System.Int32>


## Newtonsoft 0


0000   7b 7d                                               {}

## Cameronism.Json 0


0000   7b 7d                                               {}

### Equal: True

## Newtonsoft 1


0000   7b 22 30 22 3a 30 7d                                {"0":0}

## Cameronism.Json 1


0000   7b 22 30 22 3a 30 7d                                {"0":0}

### Equal: True

## Newtonsoft 2


0000   7b 22 30 22 3a 30 2c 22   31 22 3a 31 7d            {"0":0," 1":1}

## Cameronism.Json 2


0000   7b 22 30 22 3a 30 2c 22   31 22 3a 31 7d            {"0":0," 1":1}

### Equal: True

## Newtonsoft 3


0000   6e 75 6c 6c                                         null

## Cameronism.Json 3


0000   6e 75 6c 6c                                         null

### Equal: True

# System.Collections.Generic.Dictionary<System.Char, System.Int32>


## Newtonsoft 0


0000   7b 7d                                               {}

## Cameronism.Json 0


0000   7b 7d                                               {}

### Equal: True

## Newtonsoft 1


0000   7b 22 30 22 3a 30 7d                                {"0":0}

## Cameronism.Json 1


0000   7b 22 30 22 3a 30 7d                                {"0":0}

### Equal: True

## Newtonsoft 2


0000   7b 22 30 22 3a 30 2c 22   31 22 3a 31 7d            {"0":0," 1":1}

## Cameronism.Json 2


0000   7b 22 30 22 3a 30 2c 22   31 22 3a 31 7d            {"0":0," 1":1}

### Equal: True

## Newtonsoft 3


0000   6e 75 6c 6c                                         null

## Cameronism.Json 3


0000   6e 75 6c 6c                                         null

### Equal: True

# System.Collections.Generic.Dictionary<System.Guid, System.Int32[]>


## Newtonsoft 0


0000   7b 7d                                               {}

## Cameronism.Json 0


0000   7b 7d                                               {}

### Equal: True

## Newtonsoft 1


0000   7b 22 30 30 30 30 30 30   30 30 2d 30 30 30 30 2d   {"000000 00-0000-
0010   30 30 30 30 2d 30 30 30   30 2d 30 30 30 30 30 30   0000-000 0-000000
0020   30 30 30 30 30 30 22 3a   6e 75 6c 6c 7d            000000": null}

## Cameronism.Json 1


0000   7b 22 30 30 30 30 30 30   30 30 2d 30 30 30 30 2d   {"000000 00-0000-
0010   30 30 30 30 2d 30 30 30   30 2d 30 30 30 30 30 30   0000-000 0-000000
0020   30 30 30 30 30 30 22 3a   6e 75 6c 6c 7d            000000": null}

### Equal: True

## Newtonsoft 2


0000   7b 22 30 30 30 30 30 30   30 30 2d 30 30 30 30 2d   {"000000 00-0000-
0010   30 30 30 30 2d 30 30 30   30 2d 30 30 30 30 30 30   0000-000 0-000000
0020   30 30 30 30 30 30 22 3a   5b 30 5d 2c 22 38 33 38   000000": [0],"838
0030   66 35 63 30 33 2d 65 64   35 36 2d 34 31 65 31 2d   f5c03-ed 56-41e1-
0040   39 32 36 34 2d 36 38 36   62 62 31 38 64 30 61 37   9264-686 bb18d0a7
0050   37 22 3a 5b 31 2c 32 5d   7d                        7":[1,2] }

## Cameronism.Json 2


0000   7b 22 30 30 30 30 30 30   30 30 2d 30 30 30 30 2d   {"000000 00-0000-
0010   30 30 30 30 2d 30 30 30   30 2d 30 30 30 30 30 30   0000-000 0-000000
0020   30 30 30 30 30 30 22 3a   5b 30 5d 2c 22 38 33 38   000000": [0],"838
0030   66 35 63 30 33 2d 65 64   35 36 2d 34 31 65 31 2d   f5c03-ed 56-41e1-
0040   39 32 36 34 2d 36 38 36   62 62 31 38 64 30 61 37   9264-686 bb18d0a7
0050   37 22 3a 5b 31 2c 32 5d   7d                        7":[1,2] }

### Equal: True

## Newtonsoft 3


0000   6e 75 6c 6c                                         null

## Cameronism.Json 3


0000   6e 75 6c 6c                                         null

### Equal: True